### ISSUE TITLE
observation-dependent rewards

### ISSUE THREAD

**Issue by zsunberg** [JuliaPOMDP/POMDPs.jl#236]
A fairly-common request among users is to allow a relationship between observations and rewards. My initial inclination is that the best way to do this would be to add
```julia
reward(s, a, sp, o)
```
We should consider this when designing the next version.

This is, of course a similar situation to R(s, a, s') in that decision making can be done with R(s, a) if it is the expectation of R(s, a, s'), but in some cases, this expectation may be difficult to calculate, and users may not wish to augment the state. (Of course another question is "is there ever a case where R(s, a, s', o) is needed where augmenting the state space is detrimental?". Since most of these models will be generative anyways, maybe the answer is no.)

**Comment by zsunberg** [comment]
Leaning pretty strongly towards Yes on this one. We should allow defining an observation-dependent reward function.

**Comment by lassepe** [comment]
Can you provide an example in which one would want to use this? I feel like the only examples that I can think of are cases where there is deterministic observation (e.g. thinking of the Roomba problem where there is a reward for collisions). In these cases the reward can be computed from the state so having `o` in the call to `reward` would be redundant. In cases where there is observation noise, does it even make sense to have an obsevation-dependent reward?

**Comment by rejuvyesh** [comment]
I am not sure. In a sense it's a question of where do rewards come from. If we classify them into being extrinsic and intrinsic, then the extrinsic rewards are a function of (`s, a, sp`) while intrinsic rewards would be a function of (`o`). Although we could disagree and say that extrinsic rewards are a function of (`s, sp`) while intrinsic rewards are a function of (`a, o`). I'm all for allowing new ideas, but I don't think the POMDP interface allows for observation dependent rewards. It would just be an optimization hack to not have to recompute the observation (or the actual expected reward) instead.

**Comment by lassepe** [comment]
Well then, since it's useful for some applications we should provide it.

**Comment by zsunberg** [comment]
> Can you provide an example in which one would want to use this?

In medical screening for e.g. cancer, an outcome of a biopsy can cause negative psychological stress even if the disease isn't really there.

> It would just be an optimization hack to not have to recompute the observation (or the actual expected reward) instead.

I don't see it so much as an optimization hack as for allowing *clearer representation*. `o`-dependent rewards are exactly the same situation as `sp`-dependent rewards. For decision-making, it is sufficient to only have `reward(m, s, a)` which is the expected reward over all `sp` and `o`. However, if you are running a simulation, you want to be able to see that certain rewards are connected to specific next states and observations.

Bottom line - this is absolutely *not* necessary (in the same sense that Julia is not necessary because we have C ;) ), but I think it will be beneficial for communication.

**Comment by rejuvyesh** [comment]
Sure but it's a design decision for both problem writers and solver writers. How would solvers support that reward definition?

**Comment by lassepe** [comment]
I think this is only relevant for the problem writer (and for this person it is a feature rather than an annoying decision to make). Problem writers will implement `reward(s, a, sp, o)`. If the reward in their problem does not depend on `o` they simply ignore it in the implementation of `reward`. In the same way as currently people can choose to ignore `sp` if their reward function does not depend on it. Solvers should just always use the most general version (`reward(s, a, sp, o)`). Or are there POMDP solvers that don't generate an observation and therefore can not call this?

A drawback might be that the `reward` function for `MDP`s will be different from the ones used in `POMDP`s in this case, which might be a bit annoying in some cases.

**Comment by lassepe** [comment]
Well QMDPs are one example of a POMDP solver that does not generate observations. But people simply should see an error if they try to use such a solver with problem that has observation-dependent rewards (and should be notified in the requirements for the solver).

**Comment by rejuvyesh** [comment]
> However, if you are running a simulation, you want to be able to see that certain rewards are connected to specific next states and observations.

I believe this can be solved in a different way. Say with callbacks?
Main reason I am pushing back is because as a solver writer it just seems weird that the reward function depends on the observation and is another thing docs have to explain to someone coming from a general POMDP background. I understand that number of solver writers << number of problem writers, but still, it's useful to have clean simple APIs that a domain expert would find intuitive from the get go.

**Comment by MaximeBouton** [comment]
I agree with @rejuvyesh on this. The observation dependent reward could be resolved by augmenting the state which I think is the right thing to do to fit the POMDP framework.

**Comment by zsunberg** [comment]
All very good points! :) I still think we should have it and will try to respond tonight.

**Comment by zsunberg** [comment]
> Sure but it's a design decision for both problem writers and solver writers. How would solvers support that reward definition?

Yes, it is absolutely important to think about solver writers here!
- For solvers that use the generative interface, nothing will change.
- For solvers that use the explicit interface, there are a few options:
    1. Use `reward(m, s, a, sp, o)` directly (this will automatically work with problems that implement only `reward(m, s, a, sp)` or `reward(m, s, a)`)
    2. Precalculate R(s, a) by taking the expectation over `reward(m, s, a, sp, o)` and use that (`POMDPModelTools.SparseTabularPOMDP` will be able to do this automatically).
    3. Just use `reward(m, s, a)` directly, and require problem writers to implement it to use that solver. (This is the least friendly, but it is already the case for e.g. SARSOP, and having a few solvers that only tolerate `reward(m, s, a)` is better than having an entire framework that doesn't acknowledge `sp` or `o` based rewards)

> I believe this can be solved in a different way. Say with callbacks?

This isn't immediately obvious to me. Can you elaborate? (maybe link to a gist)

> it's useful to have clean simple APIs that a domain expert would find intuitive from the get go.

I have found that different people have different conceptions of what a POMDP can be. Many do not think the reward can depend on `sp`, but that is actually easy to handle. The case is the same with `o`. In my opinion, the vital criterion is that the Markov property holds for the problem so that Bellman's equation works, and that will be preserved if the rewards depend on `o`.

In the health field, problems with observation-dependent rewards are often called "POMDPs". In the past, at least one researcher has decided to write his own solver because POMDPs.jl did not support observation-dependent rewards. I don't want people to go through that unnecessary effort and do their experiments without the best solvers.

> The observation dependent reward could be resolved by augmenting the state which I think is the right thing to do to fit the POMDP framework.

I do not think that would be the right thing to do. That could easily make a problem go from a solveable size to an unsolveable size. The right thing to do if observation-dependent rewards are not allowed is to implement `reward(m, s, a)` that is the expectation over `sp` and `o`. This technique could be *automated* if we add `reward(m, s, a, sp, o)`.

**Comment by lassepe** [comment]
Viewing this as a tool for "automating the marginalization over `sp` and `o`" sounds convincing to me. As this is an interface extensions, should this be part of POMDPs.jl (to have everything at one place so people find it) or is this rather something for POMDPModelTools.jl?

**Comment by zsunberg** [comment]
Yeah I think we should keep it in POMDPs so that `gen` can use it easily

**Comment by MaximeBouton** [comment]
I need a couple clarification to be further convinced: 
-  if your problem only has state dependent rewards, do you still need to implement `reward(s, a, sp, o) = reward(s, a, sp)` ?
- would an existing solver that uses `reward(s, a, sp)` be updated to use `reward(s, a, sp, o)` instead?

**Comment by zsunberg** [comment]
> if your problem only has state dependent rewards, do you still need to implement reward(s, a, sp, o) = reward(s, a, sp) ?

No, if you implement `reward(m, s, a)`, `reward(m, s, a, sp)` and `reward(m, s, a, sp, o)` will be callable and default to `reward(m, s, a)`. Similarly, if you implement `reward(m, s, a, sp)`, `reward(m, s, a, sp, o)` will be callable and will default to `reward(m, s, a, sp)`. Problem writers should implement the version with the *fewest* number of arguments that describes their problem; solver writers should use the version with the most arguments.

> would an existing solver that uses reward(s, a, sp) be updated to use reward(s, a, sp, o) instead?

Ideally, yes, but I think what may end up happening is we will write a convenience struct in `POMDPModelTools` so that they can call something like
```julia
er = ExpectedReward(pomdp)
```
at the beginning, and then call
```julia
er(s, a, sp)
```
or
```julia
er(s, a)
```
in the inner loops to get the (precomputed or cached) expected reward for states and actions or states, actions and observations.

**Comment by zsunberg** [comment]
with the introduction of the dbn trait #256 , this issue really boils down to which dbn structure we set as default. Users can opt into or out of `o` and `sp` - dependent rewards by changing the dbn structure (see also #262 ). I still think we should make the default have the :o->:r connection, but I would be open to reconsidering.

**Comment by lassepe** [comment]
I think we should keep both the `o->r` and `sp->r` edges in the default DBN topology and document well why this may be useful. The same documentation can point to the relevant issues/discussions and maybe the DBN/gen documentation so that people understand that they are not limited to these nodes in the DBN.

**Comment by zsunberg** [comment]
This is implemented in #258

**Comment by zsunberg** [comment]
A documentation challenge with this is that QMDP must have a state-dependent reward.
